### ISSUE TITLE
scope error on ActiveRecord::Relation already defined instance method error

### ISSUE THREAD

**Issue by king601** [rails/rails#34250]
(feel free to re-title this, not sure how to describe this!)

TLDR: `You tried to define a scope named "reverse" on the model "Comment", but ActiveRecord::Relation already defined an instance method with the same name. (ArgumentError)` error is being raised in Rails 5.2+ (even on master) when previously on versions less than 5.2 it worked flawlessly, with scope names such as `reverse` or `model`

### Steps to reproduce

Run this test case:

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.2.1"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with minitest 4.
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  # These two scopes below are what is causing unexpected behavior when using 5.2 or greater
  # For our application, the scopes are limiting by type, but that's not really important, they can filter off anything
  scope :reverse, -> { where.not(post_id: nil) }
  scope :model, -> { where.not(post_id: nil) }
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.comments << Comment.create!

    assert_equal 1, post.comments.count
    assert_equal 1, Comment.count
    assert_equal post.id, Comment.first.post.id
  end
end
```

To replicate the previous expected behavior pre Rails 5.2 you can modify the script to call `gem "activerecord", "5.1.6"`

The contents inside the scope can be anything you want that's a valid scope, our application uses these two scopes to limit them by Type for the STI tables we have (like History::Model or Order::Reverse for our case)

### Expected behavior
When defining a scope named `reverse` or `model` it should successfully define the scope without an error.

### Actual behavior
When running the same code that works in Rails 5.1.6 and below, it errors out with the following stacktrace:

```
Traceback (most recent call last):
	2: from rails_scope.rb:44:in `<main>'
	1: from rails_scope.rb:45:in `<class:Comment>'
/Users/andrewfomera/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-5.2.1/lib/active_record/scoping/named.rb:175:in `scope': You tried to define a scope named "reverse" on the model "Comment", but ActiveRecord::Relation already defined an instance method with the same name. (ArgumentError)
```

This causes both the Rails server, console and tests to fail when loading the models. 

### System configuration
**Rails version**: 5.2.1 but it also fails on master branch, and 5.2.0

**Ruby version**: 2.5.1

---

If this _is_ expected or known behavior, apologizes for opening this! I believe it may be similarly related to the issue https://github.com/rails/rails/issues/34098 but I could be way off. Once it went into Rails I ran into issues trying to debug where `.reverse` was being defined at, because when I look up the methods I do not see one named reverse defined. 

So if it's now expected behavior, feel free to close and I'll modify how our apps work instead.. I opened it up because we've had these scopes here in our app for close to 5 years.

Out of our large apps, only these two scopes named `reverse` and `model` failed, I'm not sure if there are others that will trigger this issue. 

Thank you for reading and all the hard work everyone contributes, it's what makes Rails great ❤️

**Comment by matthewd** [comment]
Thanks for the detailed report!

This is the expected behaviour.

The conflicting definition is here: https://github.com/rails/rails/blob/a52c6989a0aacc50b35891265a0b5631e1ec2370/activerecord/lib/active_record/relation/delegation.rb#L71-L74

When you define a scope / class method in a model class, that method is accessible on derived relations, so you can build conditions:

```ruby
class Comment < ActiveRecord::Base
  scope :good, -> { where.not(post_id: nil) }
end

Comment.where("id > 5").good
```

`reverse` is already defined on Relation, by the above delegation to the underlying array:

```ruby
Comment.order(:name).reverse # => [#<Comment name="b">, #<Comment name="a">]
```

.. so things can get confusing/ambiguous if you define your own version. (Particularly, because we have to make the methods work on two separate classes, there can be cases where you'd get "your" `reverse` and other cases where you'd get Array's.)

The fact we enforce this is new, partly because it's caused people confusion, and partly because internal changes may have shifted the balance of which method shows up where.

Our advice is to rename your scope to something that doesn't collide.

**Comment by king601** [comment]
Thank you matthew!

**Comment by fbukevin** [comment]
I got the same issue with creating scope named 'records'.

```
ArgumentError: You tried to define a scope named "records" on the model "Record", but ActiveRecord::Relation already defined an instance method with the same name. 
``

**Comment by laffy** [comment]
I have just received the same error with a scope named `excluding` upgrading to rails `6.0.0.rc2`

This reserved scope hasn't been added to rails 6:

https://github.com/rails/rails/blob/v6.0.0.rc2/activerecord/lib/active_record/relation/delegation.rb#L85-L88

**Comment by brendon** [comment]
I got this for `without` which is mixed in from `Enumerable`.
